src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesClientProvider.java

```diff
@@ -0,0 +1,74 @@
+package org.csanchez.jenkins.plugins.kubernetes;
+
+import com.google.common.base.Objects;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import io.fabric8.kubernetes.client.KubernetesClient;
+
+import javax.annotation.CheckForNull;
+import java.io.IOException;
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+import java.security.UnrecoverableKeyException;
+import java.security.cert.CertificateEncodingException;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Manages the Kubernetes client creation per cloud
+ */
+final class KubernetesClientProvider {
+
+    private static final Cache<String, Client> clients;
+
+    private static final Integer CACHE_SIZE;
+    private static final Integer CACHE_TTL;
+
+    static {
+        CACHE_SIZE = Integer.getInteger("org.csanchez.jenkins.plugins.kubernetes.clients.cacheSize", 10);
+        CACHE_TTL = Integer.getInteger("org.csanchez.jenkins.plugins.kubernetes.clients.cacheTtl", 60);
+        clients = CacheBuilder.newBuilder()
+                .maximumSize(CACHE_SIZE)
+                .expireAfterWrite(CACHE_TTL, TimeUnit.MINUTES)
```

Comment:
Could you add a listener on the cache with `removalListener()` to close all connections hold by a client when the client is evicted from the cache? The listener would basically just call `KubernetesDefaultClient.close()`

see https://google.github.io/guava/releases/16.0/api/docs/com/google/common/cache/CacheBuilder.html#removalListener(com.google.common.cache.RemovalListener) and https://github.com/fabric8io/kubernetes-client/blob/master/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/BaseClient.java#L76

Reply:
This was address by the next commit